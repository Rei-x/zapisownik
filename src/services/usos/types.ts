export enum Frequency {
	EVERY = 'każd',
	ODD = 'nieparzyst',
	EVEN = 'parzyst',
	NEVER = 'NEVER'
}

export enum Day {
	MONDAY = 'poniedzi',
	TUESDAY = 'wtor',
	WEDNESDAY = 'środ',
	THURSDAY = 'czwart',
	FRIDAY = 'piąt',
	SATURDAY = 'sobot',
	SUNDAY = 'niedziel',
	NEVER = 'NEVER'
}

export enum LessonType {
	LECTURE = 'wyklad',
	EXERCISES = 'cwiczenia',
	LABORATORY = 'laboratorium',
	PROJECT = 'projekt',
	SEMINAR = 'seminarium'
}
