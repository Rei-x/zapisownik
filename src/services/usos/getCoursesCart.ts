export type GetCoursesCart = TypeholeRootWrapper[];
interface TypeholeRootWrapper {
	course: Course;
	term: Term;
	user_registration_status: string;
	is_registration_valid: boolean;
	limits: null;
	is_linkage_required: boolean | null;
	registrations_count: null | number;
}
interface Term {
	id: string;
	name: Name;
	date_from: string;
	date_to: string;
	order_key: number;
	is_active: boolean;
}
interface Course {
	id: string;
	name: Name;
}
interface Name {
	pl: string;
	en: string;
}
