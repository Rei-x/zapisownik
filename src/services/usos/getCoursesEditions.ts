export interface GetCoursesEditions {
	course_id?: string;
	course_name?: Coursename;
	term_id?: string;
	user_groups?: Usergroup[];
	course_units_ids?: string[];
	attributes?: unknown[];
}
interface Usergroup {
	course_unit_id: string;
	group_number: number;
	class_type: Coursename;
	class_type_id: string;
	group_url: null;
	course_id: string;
	course_name: Coursename;
	course_homepage_url: null;
	course_profile_url: string;
	course_is_currently_conducted: number;
	course_fac_id: string;
	course_lang_id: string;
	term_id: string;
	lecturers: Lecturer[];
	participants: Lecturer[];
}
interface Lecturer {
	id: string;
	first_name: string;
	last_name: string;
}
interface Coursename {
	pl: string;
	en: string;
}
