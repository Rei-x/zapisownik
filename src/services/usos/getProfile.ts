export interface GetProfile {
	id: string;
	last_name: string;
	first_name: string;
	student_status: number;
	staff_status: number;
	homepage_url: null;
	sex: string;
	email: null;
	student_number: string;
	photo_urls: Photourls;
}
interface Photourls {
	'50x50': string;
}
