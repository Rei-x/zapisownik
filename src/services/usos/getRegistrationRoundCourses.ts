export type GetRegistrationRoundCourses = TypeholeRootWrapper[];
interface TypeholeRootWrapper {
	course: Course;
	term_id: string;
	status: string;
	limits: null;
	is_linkage_required: boolean;
	registrations_count: number;
}
interface Course {
	id: string;
	name: Name;
}
interface Name {
	pl: string;
	en: string;
}
