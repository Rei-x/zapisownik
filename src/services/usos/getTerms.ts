export type GetTerms = TypeholeRootWrapper[];
interface TypeholeRootWrapper {
	id: string;
	name: Name;
	start_date: string;
	end_date: string;
	order_key: number;
	finish_date: string;
	is_active: boolean;
}
interface Name {
	pl: string;
	en: string;
}
