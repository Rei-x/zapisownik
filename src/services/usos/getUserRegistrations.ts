export type GetUserRegistrations = TypeholeRootWrapper[];
interface TypeholeRootWrapper {
	id: string;
	description: Description;
	message: Description;
	type: string;
	status: string;
	is_linkage_required: boolean;
	www_instance: string;
	faculty: Faculty;
	rounds: Round[];
	related_courses: Relatedcourse[];
}
interface Relatedcourse {
	course_id: string;
	term_id: string;
	status: string;
	limits: null;
}
interface Round {
	id: string;
	name: Description;
	status: string;
	registration_mode: string;
	start_date: string;
	end_date: string;
}
interface Faculty {
	id: string;
	name: Description;
}
interface Description {
	pl: string;
	en: string;
}
