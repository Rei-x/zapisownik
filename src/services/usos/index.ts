import type { UsosClient } from '../../api/usosClient';
import type { GetCoursesCart } from './getCoursesCart';
import type { GetCoursesEditions } from './getCoursesEditions';

import type { GetProfile } from './getProfile';
import type { GetRegistrationRoundCourses } from './getRegistrationRoundCourses';
import type { GetUserRegistrations } from './getUserRegistrations';
import * as cheerio from 'cheerio';
import makeFetchCookie from 'fetch-cookie';
import { Frequency, Day, LessonType } from './types';

const fetchWithCookie = makeFetchCookie(fetch);

const hourToTime = (hour: string) => {
	const [hours, minutes] = hour.split(':').map(Number);
	return {
		hours,
		minutes
	};
};

type Time = {
	hours: number;
	minutes: number;
};

const calculateDifference = (start: Time, end: Time) => {
	let hours = end.hours - start.hours;
	let minutes = end.minutes - start.minutes;

	if (minutes < 0) {
		hours--;
		minutes += 60;
	}

	return {
		hours,
		minutes
	};
};

export const usosService = (usosClient: UsosClient) => {
	return {
		getProfile: async () => {
			return await usosClient.get<GetProfile>(
				'users/user?fields=id|student_number|first_name|last_name|sex|student_status|staff_status|email|photo_urls|homepage_url'
			);
		},
		getCoursesCarts: async () => {
			const data = await usosClient.get<GetCoursesCart>('registrations/courses_cart');

			return data;
		},

		getUserRegistrations: async () => {
			const data = await usosClient.get<GetUserRegistrations>(
				'registrations/user_registrations?fields=id|description|message|type|status|is_linkage_required|www_instance|faculty|rounds|related_courses'
			);

			return data;
		},
		getRegistrationRoundCourses: async (roundId: string) => {
			const data = await usosClient.get<GetRegistrationRoundCourses>(
				`registrations/registration_round_courses?registration_round_id=${roundId}`
			);

			return data;
		},
		getCourseEditions: async (courseId: string, termId: string) => {
			const data = await usosClient.get<GetCoursesEditions>(
				`courses/course_edition?course_id=${courseId}&term_id=${termId}&fields=course_id|course_name|term_id|user_groups|course_units_ids|attributes`
			);

			return data;
		},
		getCourseUnitWithGroups: async (courseUnitId: string) => {
			const data = await usosClient.get(
				`courses/course_unit?course_unit_id=${courseUnitId}&fields=class_groups[participants[id]|course_unit_id|course_unit|lecturers|description]`
			);

			return data;
		},
		getUserCourses: async () => {
			const data = await usosClient.get('courses/user');

			return data;
		},
		getClassGroupTimetable: async (courseUnitId: string, groupNumber: string) => {
			const data = await usosClient.get(
				`tt/classgroup?unit_id=${courseUnitId}&group_number=${groupNumber}&fields=type|start_time|end_time|frequency|group_number|classtype_name|course_name|sm_id`
			);

			return data;
		},
		getMeetingDate: async (meetingId: string) => {
			const data = await usosClient.get(`meetings/meeting?meeting_id=${meetingId}`);

			return data;
		},
		getGroups: async (courseId: string, term?: string) => {
			const data = await fetchWithCookie(
				`https://web.usos.pwr.edu.pl/kontroler.php?_action=katalog2/przedmioty/pokazPlanZajecPrzedmiotu&prz_kod=${courseId}&plan_division=semester&plan_format=new-ui${term ? `&cdyd_kod=${term}` : ''}`,
				{
					headers: {
						'User-Agent': 'Mozilla/5.0 (X11; Linux x86_64; rv:123.0) Gecko/20100101 Firefox/123.0',
						Accept:
							'text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,*/*;q=0.8',
						'Accept-Language': 'en-US,en;q=0.5',
						'Sec-GPC': '1',
						'Upgrade-Insecure-Requests': '1',
						'Sec-Fetch-Dest': 'document',
						'Sec-Fetch-Mode': 'navigate',
						'Sec-Fetch-Site': 'none',
						'Sec-Fetch-User': '?1',
						Pragma: 'no-cache',
						'Cache-Control': 'no-cache'
					}
				}
			).then((t) => t.text());
			const $ = cheerio.load(data);

			const entries = $('timetable-entry').toArray();

			const groups = entries.map((entry) => {
				const howOften = $(entry)
					.find('span[slot="dialog-event"]')
					.text()
					.split(',')
					.map((t) => t.trim());
				const frequencyText = howOften.find((t) => t.includes('co') || t.includes('kaÅ¼d')) ?? '';

				const frequency = (() => {
					switch (true) {
						case frequencyText.includes(Frequency.EVERY):
							return Frequency.EVERY;
						case frequencyText.includes(Frequency.ODD):
							return Frequency.ODD;
						case frequencyText.includes(Frequency.EVEN):
							return Frequency.EVEN;
						default:
							return Frequency.NEVER;
					}
				})();

				const day = (() => {
					switch (true) {
						case frequencyText.includes(Day.MONDAY):
							return Day.MONDAY;
						case frequencyText.includes(Day.TUESDAY):
							return Day.TUESDAY;
						case frequencyText.includes(Day.WEDNESDAY):
							return Day.WEDNESDAY;
						case frequencyText.includes(Day.THURSDAY):
							return Day.THURSDAY;
						case frequencyText.includes(Day.FRIDAY):
							return Day.FRIDAY;
						case frequencyText.includes(Day.SATURDAY):
							return Day.SATURDAY;
						case frequencyText.includes(Day.SUNDAY):
							return Day.SUNDAY;
						default:
							return Day.NEVER;
					}
				})();
				const person = $(entry).find('div[slot="dialog-person"] > a').text().trim();
				const hours = howOften
					.find((t) => t.includes('-'))
					?.split('-')
					.map((t) => t.trim());
				const hourStart = hours?.at(0) ?? '';
				const hourEnd = hours?.at(1) ?? '';
				const name = entry.attribs['name'];
				const nameExtended = $(entry).find('span[slot="dialog-info"]').text().trim();
				const personLink = $(entry).find('div[slot="dialog-person"] > a').attr('href') ?? '';
				const groupLink = $(entry).find('span[slot="dialog-info"] > a').attr('href') ?? '';
				const hourStartTime = hourToTime(hourStart);
				const hourEndTime = hourToTime(hourEnd);

				const type = (() => {
					switch (true) {
						case name.startsWith('C'):
							return LessonType.EXERCISES;
						case name.startsWith('L'):
							return LessonType.LABORATORY;
						case name.startsWith('P'):
							return LessonType.PROJECT;
						case name.startsWith('S'):
							return LessonType.SEMINAR;
						case name.startsWith('W'):
							return LessonType.LECTURE;
						default:
							return LessonType.LECTURE;
					}
				})();
				return {
					hourStartTime,
					hourEndTime,
					duration: calculateDifference(hourStartTime, hourEndTime),
					person,
					personLink,
					groupLink,
					day,
					courseId,
					type,
					nameExtended,
					frequency,
					name
				};
			});

			return groups;
		}
	};
};
