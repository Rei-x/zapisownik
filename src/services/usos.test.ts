import { test, describe, expect } from 'bun:test';
import { usosService } from './usos';

// access_token=WQu8UChxC8gbNkv4dc4G; access_token_secret=3BV7UPLQvSST4EpBZta9ZpvQvwY8EbwzQQKVr2JF
describe('usos', () => {
	const service = usosService({
		get: function <R = unknown>(): Promise<R> {
			throw new Error('Function not implemented.');
		},
		post: function <R = unknown>(): Promise<R> {
			throw new Error('Function not implemented.');
		}
	});
	test('getCourses', async () => {
		const course = await service.getGroups('W04IST-SI0010L');

		expect(course).toMatchSnapshot();
	});
});
