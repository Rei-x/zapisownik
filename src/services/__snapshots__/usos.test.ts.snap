// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`usos getCourses 1`] = `
[
  {
    "day": "poniedzi",
    "duration": {
      "hours": 1,
      "minutes": 45,
    },
    "frequency": "każd",
    "groupLink": "https://web.usos.pwr.edu.pl/kontroler.php?_action=katalog2/przedmioty/pokazZajecia&zaj_cyk_id=62139&gr_nr=7",
    "hourEndTime": {
      "hours": 11,
      "minutes": 0,
    },
    "hourStartTime": {
      "hours": 9,
      "minutes": 15,
    },
    "name": "L, gr. 7 (129, bud. C-6)",
    "nameExtended": "Zajęcia laboratoryjne, grupa 7",
    "person": "Natalia Piórkowska",
    "personLink": "https://web.usos.pwr.edu.pl/kontroler.php?_action=katalog2/osoby/pokazOsobe&os_id=252837",
  },
  {
    "day": "wtor",
    "duration": {
      "hours": 1,
      "minutes": 30,
    },
    "frequency": "każd",
    "groupLink": "https://web.usos.pwr.edu.pl/kontroler.php?_action=katalog2/przedmioty/pokazZajecia&zaj_cyk_id=62139&gr_nr=6",
    "hourEndTime": {
      "hours": 9,
      "minutes": 0,
    },
    "hourStartTime": {
      "hours": 7,
      "minutes": 30,
    },
    "name": "L, gr. 6 (127a, bud. D-2)",
    "nameExtended": "Zajęcia laboratoryjne, grupa 6",
    "person": "Natalia Piórkowska",
    "personLink": "https://web.usos.pwr.edu.pl/kontroler.php?_action=katalog2/osoby/pokazOsobe&os_id=252837",
  },
  {
    "day": "wtor",
    "duration": {
      "hours": 1,
      "minutes": 45,
    },
    "frequency": "każd",
    "groupLink": "https://web.usos.pwr.edu.pl/kontroler.php?_action=katalog2/przedmioty/pokazZajecia&zaj_cyk_id=62139&gr_nr=1",
    "hourEndTime": {
      "hours": 11,
      "minutes": 0,
    },
    "hourStartTime": {
      "hours": 9,
      "minutes": 15,
    },
    "name": "L, gr. 1 (333c, bud. D-2)",
    "nameExtended": "Zajęcia laboratoryjne, grupa 1",
    "person": "Iuliia Arkhangelska",
    "personLink": "https://web.usos.pwr.edu.pl/kontroler.php?_action=katalog2/osoby/pokazOsobe&os_id=1755",
  },
  {
    "day": "wtor",
    "duration": {
      "hours": 1,
      "minutes": 45,
    },
    "frequency": "każd",
    "groupLink": "https://web.usos.pwr.edu.pl/kontroler.php?_action=katalog2/przedmioty/pokazZajecia&zaj_cyk_id=62139&gr_nr=2",
    "hourEndTime": {
      "hours": 13,
      "minutes": 0,
    },
    "hourStartTime": {
      "hours": 11,
      "minutes": 15,
    },
    "name": "L, gr. 2 (333c, bud. D-2)",
    "nameExtended": "Zajęcia laboratoryjne, grupa 2",
    "person": "Iuliia Arkhangelska",
    "personLink": "https://web.usos.pwr.edu.pl/kontroler.php?_action=katalog2/osoby/pokazOsobe&os_id=1755",
  },
  {
    "day": "wtor",
    "duration": {
      "hours": 1,
      "minutes": 40,
    },
    "frequency": "każd",
    "groupLink": "https://web.usos.pwr.edu.pl/kontroler.php?_action=katalog2/przedmioty/pokazZajecia&zaj_cyk_id=62139&gr_nr=8",
    "hourEndTime": {
      "hours": 16,
      "minutes": 55,
    },
    "hourStartTime": {
      "hours": 15,
      "minutes": 15,
    },
    "name": "L, gr. 8 (127b, bud. D-2)",
    "nameExtended": "Zajęcia laboratoryjne, grupa 8",
    "person": "Iuliia Arkhangelska",
    "personLink": "https://web.usos.pwr.edu.pl/kontroler.php?_action=katalog2/osoby/pokazOsobe&os_id=1755",
  },
  {
    "day": "środ",
    "duration": {
      "hours": 1,
      "minutes": 40,
    },
    "frequency": "każd",
    "groupLink": "https://web.usos.pwr.edu.pl/kontroler.php?_action=katalog2/przedmioty/pokazZajecia&zaj_cyk_id=62139&gr_nr=3",
    "hourEndTime": {
      "hours": 20,
      "minutes": 35,
    },
    "hourStartTime": {
      "hours": 18,
      "minutes": 55,
    },
    "name": "L, gr. 3 (4.44, bud. B-4)",
    "nameExtended": "Zajęcia laboratoryjne, grupa 3",
    "person": "Marcin Jodłowiec",
    "personLink": "https://web.usos.pwr.edu.pl/kontroler.php?_action=katalog2/osoby/pokazOsobe&os_id=418832",
  },
  {
    "day": "czwart",
    "duration": {
      "hours": 1,
      "minutes": 30,
    },
    "frequency": "każd",
    "groupLink": "https://web.usos.pwr.edu.pl/kontroler.php?_action=katalog2/przedmioty/pokazZajecia&zaj_cyk_id=62139&gr_nr=4",
    "hourEndTime": {
      "hours": 18,
      "minutes": 45,
    },
    "hourStartTime": {
      "hours": 17,
      "minutes": 15,
    },
    "name": "L, gr. 4 (333b, bud. D-2)",
    "nameExtended": "Zajęcia laboratoryjne, grupa 4",
    "person": "Natalia Piórkowska",
    "personLink": "https://web.usos.pwr.edu.pl/kontroler.php?_action=katalog2/osoby/pokazOsobe&os_id=252837",
  },
  {
    "day": "czwart",
    "duration": {
      "hours": 1,
      "minutes": 40,
    },
    "frequency": "każd",
    "groupLink": "https://web.usos.pwr.edu.pl/kontroler.php?_action=katalog2/przedmioty/pokazZajecia&zaj_cyk_id=62139&gr_nr=5",
    "hourEndTime": {
      "hours": 20,
      "minutes": 35,
    },
    "hourStartTime": {
      "hours": 18,
      "minutes": 55,
    },
    "name": "L, gr. 5 (333b, bud. D-2)",
    "nameExtended": "Zajęcia laboratoryjne, grupa 5",
    "person": "Natalia Piórkowska",
    "personLink": "https://web.usos.pwr.edu.pl/kontroler.php?_action=katalog2/osoby/pokazOsobe&os_id=252837",
  },
]
`;
